Copy the gas model from wiki to a md doc to allow pull requests.

virgil-serbanuta opened the pull request.

dwightguth commented on gas.md:
````diff
@@ -0,0 +1,554 @@
+* `JUMPI`
+    ```hs
+    computationCost(JUMPI(_) cREG)
+      | not (isZero cREG) = bitCost + jumpCost
````
also it seems like for simplicity's sake we should just pay the upper bound of these two costs in both cases

virgil-serbanuta replied:
Gmid (JUMP) = 8, Ghigh (JUMPI) = 10.

bitCost seems to be (in c++/Java terms) the cost of checking whether the 'length' field of a register is 0, which is usually 1-2 native instructions (assuming that we compile Iele to native code), which, I assume, will mean a cost of 1-2 on the same scale as above. The exact cost depends on the actual IELE register representation in actual machine terms, but 2 seems reasonable.

dwightguth commented on gas.md:
```diff
@@ -0,0 +1,554 @@
+  computationCost(LOG0 rMEMSTART rMEMWIDTH) =
+    logCost + value rMEMWIDTH * logDataCost
+  computationCost(LOG1 rMEMSTART rMEMWIDTH rW0) =
+    logCost + value rMEMWIDTH * logDataCost + registerSize rW0 * logTopicWordCost
```
do we actually want unbounded log topics? I'm unsure.

dwightguth replied:
Note that we bound memory indices by 2^256 so there is precedent here

virgil-serbanuta replied:
I'm not sure we want unbounded log topics either. My first guess is that they have potential to create some bad system issues, so we don't want them. However,  from a gas computation perspective, this should not be relevant. If rWi are limited in size in the IELE semantics, I could use the same formula or I could put a flat fee here, as EVM does.

bmmoore commented:
What is going on with this pull request? We now have a `gas.md` in the `gas` branch in this repository, so I think at least the source or target branch is now wrong.

Have the comments been addressed on the gas branch?

virgil-serbanuta commented:
I think I have addressed the comments on the gas branch.

I originally intended to put everything on the 'develop' branch, so I made a pull request with Traian's version here. However, Dwight didn't answer until today, so I created another 'gas' branch to have incremental code reviews.

Since Dwight started reviewing this, I would continue in parallel with both pull requests, then, once this PR is submitted, I would merge everything into the 'develop branch'.

bmmoore commented on gas.md:
````diff
@@ -0,0 +1,554 @@
+    ```hs
+    computationCost(EXP rREG wREG1 wREG2)
+      | isZero wREG2 = bitCost + wordCopyCost
+      | otherwise    = constExp * wordCost * (0.5 * approxKara l * (approxKara e - 1) + 2 * l * (e - 1)))
````
For exponentiation a reasonable estimate needs to use the value of `e`. For non-modular exponentiation we could maybe just immediately report out of gas if `e` is above some maximum length and not worry about the variable size.

For `EXPMOD` much larger exponents may be tractable. We may just have to do a two-phase gas check, first testing there is enough gas to evaluate a formula involving `e` and then doing the full cost calculation.

virgil-serbanuta replied:
I'm not sure I understand what happens here, so I'll wait a bit to see if Traian will answer. From the general shape of the formula, I guess that one should have ((registersize e)-1) instead of (e-1), but can't tell for sure right now. The same below.

traiansf replied:
I agree with @bmmoore that we could do a two-phase check, and also include the price for the checks in the overall gas costs.

dwightguth replied:
I would take a look at what EVM did. They had a way of computing the gas cost of the exponent based on measuring the bit width of the exponent and doing bitwise arithmetic on the lower-order 32 bytes of the exponent. I think that if we can do something like that, it will be much better than trying to do arbitrary precision arithmetic, even though it may make the gas cost higher.

virgil-serbanuta commented on gas.md:
````diff
@@ -0,0 +1,554 @@
+    ```hs
+    estimatedResultSize(EXPMOD rREG wREG1 wREG2)
+      | isZero wREG2 = 1
+      | otherwise    = l3
````
I'll fix this issue. However, there is a question which is important for my other pull requests:

I have no idea if Traian intended this to actually be 100% valid Haskell code. My working assumption was that this was pseudocode using Haskell-like syntax.

Should I attempt to use 100% valid Haskell syntax in my other pull requests?

virgil-serbanuta commented:
If it wasn't obvious, this PR is Traian's work and I just put it into a form better suited for reviews.

I think I have addressed all issues except one. Would it be fine to just add a TODO there?
